tiledb/sm/tile/generic_tile_io.cc

```diff
@@ -119,13 +120,13 @@ tuple<Status, optional<Tile>> GenericTileIO::read_generic(
 
 tuple<Status, optional<GenericTileIO::GenericTileHeader>>
 GenericTileIO::read_generic_tile_header(
-    const StorageManager* sm, const URI& uri, uint64_t file_offset) {
+    VFS* vfs, const URI& uri, uint64_t file_offset) {
   GenericTileHeader header;
 
   // Read the fixed-sized part of the header from file
   tdb_unique_ptr<Buffer> header_buff(tdb_new(Buffer));
   RETURN_NOT_OK_TUPLE(
-      sm->read(
+      vfs->read(
```

Comment:
The Buffer class is used in too many places... Let's take the opportunity to get rid of the Buffer class use here and use another container (maybe vector) since we know the size to read. Below we can allocate another vector once the filter pipeline size is known. This should also get rid of the extra function added to VFS.

Reply:
This `read` function is called in exactly two places from a single function `GenericTileIO::read_generic_tile_header`, where the buffer structure is used in an essential way to do stream-like input. Eliminating the buffer from this function would require entirely rewriting it.

I'm in favor of doing this, but not in this PR. What we can do in this PR is to mark the VFS function for elimination and the `GenericTileIO` function for rewriting, each with comments pointing at each other.

In the future when we do this, in addition, we should move `struct GenericTileHeader` into the `storage_format` directory and begin using that directory for its intended purpose--to isolate storage formatting and all the low-level I/O attendant to it.

Reply:
I have an updated version of this PR that uses the Deserializer class to replace the use of Buffer for deserialization. The biggest thing to note here is that the `Buffer::read` method returns a Status for error while the `Derserializer::read` method throws instead. I was concerned about the change at first, however I noticed that this function already can throw an error from`Deserializer::read` when deserializing filter pipelines [1, 2] so I think it should be fine?

[1] https://github.com/TileDB-Inc/TileDB/blob/0ed11d3c591e06b956bc4c977d452b803c055113/tiledb/sm/tile/generic_tile_io.cc#L161
[2] https://github.com/TileDB-Inc/TileDB/blob/0ed11d3c591e06b956bc4c977d452b803c055113/tiledb/sm/filter/filter_pipeline.cc#L666-L667

Reply:
I should probably have linked to the change set for the new approach: https://github.com/TileDB-Inc/TileDB/compare/pd/sc23377-remove-sm-from-generic-tile-2

Reply:
> throws instead

It should be fine for it to throw. We're all set up at this point that such changes aren't problematic. Where there's even a problem, it's been with unit tests that expect non-throwing behavior.

Since you're rewriting this function, it would be best to change its signature as well and drop the (transitional) tuple-optional pattern and just return `GenericTileHeader`. The function is only called in two places, so this is an easy change.